ui/src/app/utils/errors.tsx

```diff
@@ -85,3 +90,134 @@ export async function fetchWithSystemErrorHandler<T>(
+ * @param expectedResponseMatcher An optional handler for errors which should be considered successes
+ * @param customErrorResponseFormatter An optional handler for expected responses; if missing or this handler returns
+ * undefined, the default formatter will be used.
+ * @returns A NotificationStore object suitable for an error modal, if appropriate; undefined otherwise
+ */
+export const errorHandlerWithFallback = (
+  anyApiErrorResponse: any,
+  expectedResponseMatcher?: (ErrorResponse) => boolean,
+  customErrorResponseFormatter?: (ErrorResponse) => NotificationStore
+): Promise<NotificationStore> =>
+  convertAPIError(anyApiErrorResponse).then((errorResponse) => {
+    // if this "error" is expected and should instead be considered a success
+    if (expectedResponseMatcher?.(errorResponse)) {
+      return undefined;
+    }
+
+    // the custom error response for this error, if applicable
+    return (
+      customErrorResponseFormatter?.(errorResponse) || {
+        title: FALLBACK_ERROR_TITLE,
+        message: defaultErrorResponseFormatter(errorResponse),
+      }
+    );
+  });
+
+interface FetchOptions<T, U> {
+  onSuccess?: (t: T) => U;
+  expectedResponseMatcher?: (ErrorResponse) => boolean;
+  customErrorResponseFormatter?: (ErrorResponse) => NotificationStore;
+}
+/**
+ * Call an (async) API function and optionally execute another function if successful.  If not, pop up a modal with a
+ * suitable message by setting the notificationStore (see NotificationModal for more details).
+ *
+ * The caller may supply a matcher for expected error responses which should be considered successes, and a custom
+ * formatter for expected failure responses; if neither of these match, the defaultErrorResponseFormatter will be used.
+ *
+ * Example: an API call getInfo() sometimes appropriately returns 404s which should not be considered errors,
+ * and there's a common error case (HTTP 429, Error Code TOO_FAST) which we want to handle specially.
+ *
+ * fetchWithErrorModal(
+ *  () => getInfo(),
+ *  {
+ *    onSuccess: (ir: InfoResult) => handleInfoResult(ir),
+ *    expectedResponseMatcher: (er: ErrorResponse) => er.statusCode === 404,
+ *    customErrorResponseFormatter: (er: ErrorResponse) =>
+ *      er.statusCode === 429 && er.errorCode === ErrorCode.TOO_FAST && {
+ *        title: 'Please try again',
+ *        message: 'The server is currently handling too many requests',
+ *      }
+ *  });
+ *
+ * @param apiCall The API function to call and handle
+ * @param options:
+ *  onSuccess The optional action to take if the API call is successful
+ *  expectedResponseMatcher An optional handler for errors which should be considered successes
+ *  customErrorResponseFormatter An optional handler for expected responses; if missing or this handler returns
+ * undefined, the default formatter will be used.
+ * @returns The result of the API function call, if successful; undefined otherwise
+ */
+export async function fetchWithErrorModal<T, U>(
```

Comment:
naming question: should I include "catch" in the name to make it clear that errors are handled here?

Specifically, callers should not expect `.then()` and `.catch()` to work as expected on the result from this function

Reply:
Maybe `withErrorModalCatch`?

Reply:
no longer applicable - the promise will be returned as-is after modal handling